File: Src/java/engine-fhir/src/test/java/org/opencds/cqf/cql/engine/fhir/searchparam/TestSearchParameterResolver.java
package org.opencds.cqf.cql.engine.fhir.searchparam;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

import ca.uhn.fhir.context.FhirContext;
import ca.uhn.fhir.context.FhirVersionEnum;
import ca.uhn.fhir.context.RuntimeSearchParam;
import ca.uhn.fhir.model.api.IQueryParameterType;
import ca.uhn.fhir.rest.api.RestSearchParameterTypeEnum;
import org.apache.commons.lang3.tuple.Pair;
import org.junit.jupiter.api.Test;

class TestSearchParameterResolver {
    @Test
    void returnsNullPathReturnsNull() {
        SearchParameterResolver resolver = new SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.DSTU3));

        RuntimeSearchParam param = resolver.getSearchParameterDefinition("Patient", null);
        assertNull(param);
    }

    @Test
    void nullDataTypeReturnsNull() {
        SearchParameterResolver resolver = new SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.DSTU3));

        RuntimeSearchParam param = resolver.getSearchParameterDefinition(null, "code");
        assertNull(param);
    }

    @Test
    void dstu3SearchParams() {
        SearchParameterResolver resolver = new SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.DSTU3));

        RuntimeSearchParam param = resolver.getSearchParameterDefinition("Patient", "id");
        assertNotNull(param);
        assertEquals("_id", param.getName());

        param = resolver.getSearchParameterDefinition(
                "MedicationAdministration", "medication", RestSearchParameterTypeEnum.TOKEN);
        assertNotNull(param);
        assertEquals("code", param.getName());

        param = resolver.getSearchParameterDefinition(
                "MedicationAdministration", "medication", RestSearchParameterTypeEnum.REFERENCE);
        assertNotNull(param);
        assertEquals("medication", param.getName());

        param = resolver.getSearchParameterDefinition("Encounter", "period");
        assertNotNull(param);
        assertEquals("date", param.getName());

        param = resolver.getSearchParameterDefinition("Encounter", "reason");
        assertNotNull(param);
        assertEquals("reason", param.getName());

        param = resolver.getSearchParameterDefinition("Encounter", "subject");
        assertNotNull(param);
        assertEquals("patient", param.getName());

        param = resolver.getSearchParameterDefinition("Encounter", "type");
        assertNotNull(param);
        assertEquals("type", param.getName());
    }

    @Test
    void dstu3DateSearchParams() {
        SearchParameterResolver resolver = new SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.DSTU3));

        RuntimeSearchParam param = resolver.getSearchParameterDefinition(
                "ProcedureRequest", "authoredOn", RestSearchParameterTypeEnum.DATE);
        assertNotNull(param);
        assertEquals("authored", param.getName());
    }

    @Test
    void r4SearchParams() {
        SearchParameterResolver resolver = new SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.R4));

        RuntimeSearchParam param = resolver.getSearchParameterDefinition("Patient", "id");
        assertNotNull(param);
        assertEquals("_id", param.getName());

        param = resolver.getSearchParameterDefinition(
                "MedicationAdministration", "medication", RestSearchParameterTypeEnum.TOKEN);
        assertNotNull(param);
        assertEquals("code", param.getName());

        param = resolver.getSearchParameterDefinition(
                "MedicationAdministration", "medication", RestSearchParameterTypeEnum.REFERENCE);
        assertNotNull(param);
        assertEquals("medication", param.getName());

        param = resolver.getSearchParameterDefinition("Encounter", "period");
        assertNotNull(param);
        assertEquals("date", param.getName());

        param = resolver.getSearchParameterDefinition("Encounter", "reasonCode");
        assertNotNull(param);
        assertEquals("reason-code", param.getName());

        param = resolver.getSearchParameterDefinition("Encounter", "subject");
        assertNotNull(param);
        assertEquals("subject", param.getName());

        param = resolver.getSearchParameterDefinition("Encounter", "type");
        assertNotNull(param);
        assertEquals("type", param.getName());

        param = resolver.getSearchParameterDefinition("Observation", "code");
        assertNotNull(param);
        assertEquals("code", param.getName());

        param = resolver.getSearchParameterDefinition("Observation", "subject");
        assertNotNull(param);
        assertEquals("subject", param.getName());
    }

    @Test
    void r4DateSearchParams() {
        SearchParameterResolver resolver = new SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.R4));

        RuntimeSearchParam param =
                resolver.getSearchParameterDefinition("ServiceRequest", "authoredOn", RestSearchParameterTypeEnum.DATE);
        assertNotNull(param);
        assertEquals("authored", param.getName());

        param = resolver.getSearchParameterDefinition("Condition", "onset", RestSearchParameterTypeEnum.DATE);
        assertNotNull(param);
        assertEquals("onset-date", param.getName());

        param = resolver.getSearchParameterDefinition("Condition", "abatement", RestSearchParameterTypeEnum.DATE);
        assertNotNull(param);
        assertEquals("abatement-date", param.getName());

        param = resolver.getSearchParameterDefinition("Observation", "effective", RestSearchParameterTypeEnum.DATE);
        assertNotNull(param);
        assertEquals("date", param.getName());

        param = resolver.getSearchParameterDefinition("Observation", "value", RestSearchParameterTypeEnum.DATE);
        assertNotNull(param);
        assertEquals("value-date", param.getName());
    }

    @Test
    void r4ReferenceParameter() {
        FhirContext context = FhirContext.forCached(FhirVersionEnum.R4);
        SearchParameterResolver resolver = new SearchParameterResolver(context);
        Pair<String, IQueryParameterType> actual =
                resolver.createSearchParameter("Patient", "Observation", "subject", "123");

        assertEquals("Patient/123", actual.getRight().getValueAsQueryToken(context));
    }

    @Test
    void r4TokenParameter() {
        FhirContext context = FhirContext.forCached(FhirVersionEnum.R4);
        SearchParameterResolver resolver = new SearchParameterResolver(context);
        Pair<String, IQueryParameterType> actual =
                resolver.createSearchParameter("Patient", "Observation", "code", "123");

        assertEquals("123", actual.getRight().getValueAsQueryToken(context));
    }
}

File: Src/java/engine-fhir/src/test/kotlin/org/opencds/cqf/cql/engine/fhir/searchparam/TestSearchParameterResolver.kt
package org.opencds.cqf.cql.engine.fhir.searchparam

import ca.uhn.fhir.context.FhirContext
import ca.uhn.fhir.context.FhirVersionEnum
import ca.uhn.fhir.rest.api.RestSearchParameterTypeEnum
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class TestSearchParameterResolver {
    @Test
    fun returnsNullPathReturnsNull() {
        val resolver = SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.DSTU3))

        val param = resolver.getSearchParameterDefinition("Patient", null)
        Assertions.assertNull(param)
    }

    @Test
    fun nullDataTypeReturnsNull() {
        val resolver = SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.DSTU3))

        val param = resolver.getSearchParameterDefinition(null, "code")
        Assertions.assertNull(param)
    }

    @Test
    fun dstu3SearchParams() {
        val resolver = SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.DSTU3))

        var param = resolver.getSearchParameterDefinition("Patient", "id")
        Assertions.assertNotNull(param)
        Assertions.assertEquals("_id", param!!.name)

        param =
            resolver.getSearchParameterDefinition(
                "MedicationAdministration",
                "medication",
                RestSearchParameterTypeEnum.TOKEN,
            )!!
        Assertions.assertNotNull(param)
        Assertions.assertEquals("code", param.name)

        param =
            resolver.getSearchParameterDefinition(
                "MedicationAdministration",
                "medication",
                RestSearchParameterTypeEnum.REFERENCE,
            )!!
        Assertions.assertEquals("medication", param.name)

        param = resolver.getSearchParameterDefinition("Encounter", "period")!!
        Assertions.assertEquals("date", param.name)

        param = resolver.getSearchParameterDefinition("Encounter", "reason")!!
        Assertions.assertEquals("reason", param.name)

        param = resolver.getSearchParameterDefinition("Encounter", "subject")!!
        Assertions.assertEquals("patient", param.name)

        param = resolver.getSearchParameterDefinition("Encounter", "type")!!
        Assertions.assertEquals("type", param.name)
    }

    @Test
    fun dstu3DateSearchParams() {
        val resolver = SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.DSTU3))

        val param =
            resolver.getSearchParameterDefinition(
                "ProcedureRequest",
                "authoredOn",
                RestSearchParameterTypeEnum.DATE,
            )
        Assertions.assertNotNull(param)
        Assertions.assertEquals("authored", param!!.name)
    }

    @Test
    fun r4SearchParams() {
        val resolver = SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.R4))

        var param = resolver.getSearchParameterDefinition("Patient", "id")
        Assertions.assertNotNull(param)
        Assertions.assertEquals("_id", param!!.name)

        param =
            resolver.getSearchParameterDefinition(
                "MedicationAdministration",
                "medication",
                RestSearchParameterTypeEnum.TOKEN,
            )!!
        Assertions.assertEquals("code", param.name)

        param =
            resolver.getSearchParameterDefinition(
                "MedicationAdministration",
                "medication",
                RestSearchParameterTypeEnum.REFERENCE,
            )!!
        Assertions.assertEquals("medication", param.name)

        param = resolver.getSearchParameterDefinition("Encounter", "period")!!
        Assertions.assertEquals("date", param.name)

        param = resolver.getSearchParameterDefinition("Encounter", "reasonCode")!!
        Assertions.assertEquals("reason-code", param.name)

        param = resolver.getSearchParameterDefinition("Encounter", "subject")!!
        Assertions.assertEquals("subject", param.name)

        param = resolver.getSearchParameterDefinition("Encounter", "type")!!
        Assertions.assertEquals("type", param.name)

        param = resolver.getSearchParameterDefinition("Observation", "code")!!
        Assertions.assertEquals("code", param.name)

        param = resolver.getSearchParameterDefinition("Observation", "subject")!!
        Assertions.assertEquals("subject", param.name)
    }

    @Test
    fun r4DateSearchParams() {
        val resolver = SearchParameterResolver(FhirContext.forCached(FhirVersionEnum.R4))

        var param =
            resolver.getSearchParameterDefinition(
                "ServiceRequest",
                "authoredOn",
                RestSearchParameterTypeEnum.DATE,
            )!!
        Assertions.assertEquals("authored", param.name)

        param =
            resolver.getSearchParameterDefinition(
                "Condition",
                "onset",
                RestSearchParameterTypeEnum.DATE,
            )!!
        Assertions.assertEquals("onset-date", param.name)

        param =
            resolver.getSearchParameterDefinition(
                "Condition",
                "abatement",
                RestSearchParameterTypeEnum.DATE,
            )!!
        Assertions.assertEquals("abatement-date", param.name)

        param =
            resolver.getSearchParameterDefinition(
                "Observation",
                "effective",
                RestSearchParameterTypeEnum.DATE,
            )!!
        Assertions.assertEquals("date", param.name)

        param =
            resolver.getSearchParameterDefinition(
                "Observation",
                "value",
                RestSearchParameterTypeEnum.DATE,
            )!!
        Assertions.assertEquals("value-date", param.name)
    }

    @Test
    fun r4ReferenceParameter() {
        val context = FhirContext.forCached(FhirVersionEnum.R4)
        val resolver = SearchParameterResolver(context)
        val actual = resolver.createSearchParameter("Patient", "Observation", "subject", "123")!!

        Assertions.assertEquals("Patient/123", actual.getRight()!!.getValueAsQueryToken(context))
    }

    @Test
    fun r4TokenParameter() {
        val context = FhirContext.forCached(FhirVersionEnum.R4)
        val resolver = SearchParameterResolver(context)
        val actual = resolver.createSearchParameter("Patient", "Observation", "code", "123")!!

        Assertions.assertEquals("123", actual.getRight()!!.getValueAsQueryToken(context))
    }
}
